pyswarms/backend/topology/pyramid.py

```diff
@@ -22,8 +22,18 @@
 
 
 class Pyramid(Topology):
-    def __init__(self):
+    def __init__(self, static):
```

Comment:
I suggest to have a default value for `static`, i.e., `static=False`.

Reply:
It was actually a very conscious choice to implement the `static` attribute it in that way. There are two reasons:
* **An active choice**: If we implement it without a default statement the user has actively to choose which topology he actually wants. There are arguments for both: *static* uses less computation time and saves some resources on the user's side, *dynamic* is the traditional way. So I chose to not influence the user by defaults (there is an interesting book about this called *Nudge*)

* **Readability**: By implementing it so the user has to choose whether the topology is static or dynamic he has more control over his code. You can look at it and immediately see that the topology is static or dynamic. This improves the readability and comprehensibility of the code.

Reply:
Hi @whzup , I understand your point about having the user check if he/she needs a static or dynamic topology. However, this requires a lot of mental processing on the user-side to differentiate between the two. If we can provide a reasonable default and it works, then it's all good. I still maintain that we set a default value for this parameter.

1. **On active choices**
> If we implement it without a default statement the user has actively to choose which topology he actually wants.

Default values provide a baseline from which a user can adapt his/her system: "If a `dynamic` topology didn't work, then maybe a `static` topology will." as opposed to asking "What's the difference between `dynamic` and `static`?" right off the bat. Sure, I can try one first randomly, but why not just set a default?

Our main philosophy is to keep things simple. We only expose the API *whenever needed*. 
The library requires a necessary input from the user, and it gives the user's expected output. All customization parameters should either be defaulted or unexposed. The `static=False` option is the most reasonable for most use-cases, so we set it that way.

We want to make the users familiar with the `input -> output` (or f(x) = y) scheme before introducing other parameters. Again, in this manner, the user checks if his system works given a baseline, then he/she adapts the algorithm until it clicks.

A notable example for this would be the [SVM classifier](http://scikit-learn.org/stable/modules/generated/sklearn.svm.SVC.html) of scikit-learn. Sensible defaults are given and it provides a baseline for experimentation (e.g. if RBF doesn't work, maybe a polynomial kernel will, etc.).

>  There are arguments for both: static uses less computation time and saves some resources on the user's side

I might need some supporting evidence for this aside from looking at the code. Static *may* be "faster," but is it significantly faster? If you can profile this that would be nice. 

2. **On readability**
>  By implementing it so the user has to choose whether the topology is static or dynamic he has more control over his code.

We have solved the "more control over his code" part when we implemented the parameter itself. Again, actively choosing something still requires a lot of thought. Let's give them a sensible default and have them iterate if it didn't work.

Reply:
> Where would you search if the code does not work? 

In the documentation, I mean seriously. 

Not  *if it does not work*, but if it does not optimize to an optima you are expecting. By that time you would probably check what can be done with the topology. The keyword is to provide **sensible** defaults. In this case our defaults will be the traditional approach, as you've mentioned. :+1: 

Also, look at how `static`is defined: `True` or `False`. Think of this as a switch with a default state (in this case, turned-off). What a user can do is to turn-on a switch so that they get out of that traditional scheme and lock themselves into a fixed set of neighbors. Ok, sure we can change this to  `connection='static'` or `'dynamic'`, but if we have a discrete choice between `1` or `2`, why not just make it `0` or `1`?

> I think in the SVM example you gave, the choice to have default values makes sense because you have to make very complex decisions about the parameters and you have a lot of parameters.

Think of `shrinking`, `probability`, or `class_weights`. I can call SVM, provide C and gamma, and care less about what shrinking does. I trust that it is a sensible default, and I just focus on building the model. If I want to further optimize, then I look at what turning it off means.

Same with `probability`, I know that by setting it to `True`, I sacrifice training speed (according to the docs), I just turn it on when I need it and focus on my model.

> The user has to ask one simple question "static or dynamic?" and that's it. I think even with the active choice of the parameter it still remains very simple as long as we don't add any more.

Again, there's a cognitive overhead. I still need to check for what `static` means or what `dynamic` does. We can provide the traditional default first, and have them change that whenever needed.

> If we strictly follow this philosophy we'd also have to set defaults for the options, bounds, swarm size etc. in the optimizers, wouldn't we?

We already have defaults for bounds. We can keep the swarm size flexible but you're correct to set a default  `options` dictionary. I've been thinking of doing that for some time, but I realized that the common workflow in PSO is that researchers choose values for the cognitive and social parameters and so we follow what they usually do.

> I was actually thinking about profiling the whole GeneralOptimizer class with every topology for v0.4.0 so we can show some examples in the README, what do you think about that? Might be a good first-timers issue as well.

This is a good idea.

> PS: I heard that in many Asian cultures there is a very strong hierarchical structure in projects and that there is little questioning about decisions of the supervisor. This is a very European thing to do. I hope I don't insult you in any way by doing so. Just want this project to become a great product 

All fine, but no need to mention this *explicitly* to be honest. Same with `Pyramid(static)` :laughing: 
We're just open-source enthusiasts, after all :+1: : 

> For exactly the simplicity you're describing we're still keeping the GlobalOptimizer and the LocalOptimizer classes.

Again, simplicity is key. We want the user to use our methods right off the bat. Customization happens *when needed* and not at every step of the way. I'm still trying to find the right balance between this philosophy and flexibility. We're flexible, sure. That's the reason why we try to implement as many options as possible. :100: 

Thanks for raising these thought-provoking questions!

Reply:
Again, the keyword is **convention over configuration**. It’s a software engineering principle, you can look that one up :+1:

Reply:
How about a compromise? Add a `logger.info` stating that you are running on a `static` or `dynamic` topology. This gives the user awareness that there is a different topological implementation outside of tradition.

```
[INFO] Running on `dynamic` topology, neighbors are updated regularly. Set `static=True` for fixed neighbors.
```

Or something like that.

Reply:
Hi @whzup ! Yes, I'm expecting that Zen of Python argument. It's nontrivial, to be honest. However,  the inverse of convention over configuration violates "Simple is better than complex." This just shows that some design principles can sit in tension with each other. Again, our principle would be **sensible defaults** and to just stick with convention unless really needed. :+1: 

Yes, the log information should be a good compromise! :100:

Reply:
Can I just add code like that to the GeneralOptimizer:

```python
if not static:
    logger.info(dynamic_message)
```

Or is there a special place where I'd have to include this? I haven't looked into the logging too much, to be honest 😄.

Reply:
Similar to that, yes. So we have `cli_print` function [here](https://github.com/ljvmiranda921/pyswarms/blob/master/pyswarms/utils/console_utils.py#L13). You can just use the function.

Yes, maybe you can put it at `GeneralOptimizer`'s `__init__` method? Or in `optimize`? Depends on where you think it's most applicable. :+1:

Reply:
Whoops, sorry because I thought of something @whzup...

What if we put it inside the __init__ method of the Topology class itself?

```python
try:
    if self.static:
        cli_print(message)
except AttributeError: # in case it does not exist such as in Star
    pass
```

With this, we don't need to put the logger in all optimizers. It's all in the base topology class

Reply:
Should I include the two attributes in the `Topology` class? I think it's a bit dirty to just leave the `Star` topology out. The `neighbor_idx` attribute could be useful for future PSO algorithms (e.g. for accessing the neighbours outside the topology). In the `Topology` class it'd look like this:

```python
class Topology(object):
    def __init__(self, static=False, **kwargs):
        """Initializes the class"""

        # Initialize logger
        self.logger = logging.getLogger(__name__)

        self.static = static
        self.neighbor_idx = None

        try:
            if self.static:
                cli_print("Running on `dynamic` topology, neighbors are updated regularly."
                          "Set `static=True` for fixed neighbors.",
                          1,
                          0,
                          self.logger)
        except AttributeError:
            pass
```

It also adds to DRY 😄.